anaconda_project/internal/plugins/registry.py

```diff
@@ -0,0 +1,204 @@
+        # implements the plugins API requirements
+        self.load_plugin()
+
+    @staticmethod
+    def is_plugin_candidate_path(path):
+        """Return True if the path is a potential plugin path."""
+        if (isdir(path) and exists(join(path, 'plugin.py'))) or path.endswith('.py'):
+            return True
+        return False
+
+    @property
+    def failed(self):
+        """Return if True if the plugin has failed to load, False otherwise.
+
+        Returns:
+
+            bool
+        """
+        return self._runner._failed
+
+    @property
+    def error(self):
+        """Return the error occurred when loading the plugin, None otherwise.
+
+        Returns:
+
+            str / None (if no error occurred)
+        """
+        return self._runner._error
+
+    @property
+    def error_detail(self):
+        """Return the error defails occurred when loading the plugin.
+        None otherwise
+
+        Returns:
+
+            str / None (if no error occurred)
+        """
+        return self._runner._error_detail
+
+    @property
+    def loggers(self):
+        """List of activate loggers."""
+        return verbose._verbose_loggers
+
+    @property
+    def ProjectCommand(self):
+        # TODO: we should add a Plugin API validation check on the plugin module we get
+        return self._module.ProjectCommand  # noqa
+
+    @staticmethod
+    def create(path):
+        """Return a plugin loaded from path.
+
+        Args:
+            path (str) : path to a file or directory for creating a Plugin.
+
+        Returns:
+            Plugin or None (if creation fails)
+        """
+        path = os.path.abspath(path)
+
+        if os.path.isdir(path):
+            plugin_class = PackagePlugin
+        else:
+            if path.endswith(".py"):
+                plugin_class = ModulePlugin
+            else:
+                log("Expected a '.py' script, got: '%s'" % path, 'error')
+
+        plugin = plugin_class(path)
+        if plugin.failed:
+            msg = "Error loading %s:\n\n%s\n%s " % (path, plugin.error, plugin.error_detail)
+            log(msg)
+            return
+
+        return plugin
+
+    def init_plugin(self):
+        """Specific plugin logic to be overwritten by customized plugin classes.
+        This method should at least init the following attributes:
+
+        - _source
+
+        """
+        raise NotImplementedError
+
+    def load_plugin(self):
+        """Run the plugin."""
+        self._runner = CodeRunner(self._source, self.path)
+        if not self._runner.failed:
+            self._module = self._runner.new_module(self.name)
+            self._runner.run(self._module)
+
+    def log(self, msg, level='info'):
+        """Log log to all activate loggers."""
+        for logger in self.loggers:
+            getattr(logger, level)(msg)
+
+
+class ModulePlugin(Plugin):
+    """Module (single file) plugin."""
+
+    def init_plugin(self):
+        with open(self.path, 'r') as f:
```

Comment:
this isn't coping with encoding I don't think?
(In most of the code I use `codecs.open(self.path, 'r', encoding='utf-8')` but not sure that's right for a python file)